backends-velox/src/main/scala/io/glutenproject/execution/VanillaColumnarToVeloxColumnarExec.scala

```diff
@@ -0,0 +1,162 @@
+
+  override protected def doValidateInternal(): ValidationResult = {
+    BackendsApiManager.getValidatorApiInstance.doSchemaValidate(schema) match {
+      case Some(reason) =>
+        ValidationResult(
+          isValid = false,
+          Some(
+            s"Input schema contains unsupported type when performing columnar" +
+              s" to columnar for $schema " + s"due to $reason"))
+      case None =>
+        ValidationResult(isValid = true, None)
+    }
+  }
+
+  override def doExecuteColumnarInternal(): RDD[ColumnarBatch] = {
+    val numInputBatches = longMetric("numInputBatches")
+    val numOutputBatches = longMetric("numOutputBatches")
+    val convertTime = longMetric("convertTime")
+    // This avoids calling `schema` in the RDD closure, so that we don't need to include the entire
+    // plan (this) in the closure.
+    val localSchema = schema
+    child.executeColumnar().mapPartitions {
+      rowIterator =>
+        VanillaColumnarToVeloxColumnarExec.toColumnarBatchIterator(
+          rowIterator,
+          localSchema,
+          numInputBatches,
+          numOutputBatches,
+          convertTime)
+    }
+  }
+
+  override protected def withNewChildInternal(newChild: SparkPlan): SparkPlan = {
+    copy(child = newChild)
+  }
+}
+
+object VanillaColumnarToVeloxColumnarExec {
+
+  def toColumnarBatchIterator(
+      it: Iterator[ColumnarBatch],
+      schema: StructType,
+      numInputBatches: SQLMetric,
+      numOutputBatches: SQLMetric,
+      convertTime: SQLMetric): Iterator[ColumnarBatch] = {
+    if (it.isEmpty) {
+      return Iterator.empty
+    }
+
+    val arrowSchema =
+      SparkArrowUtil.toArrowSchema(schema, SQLConf.get.sessionLocalTimeZone)
+    val jniWrapper = VanillaColumnarToNativeColumnarJniWrapper.create()
+    val allocator = ArrowBufferAllocators.contextInstance()
+    val cSchema = ArrowSchema.allocateNew(allocator)
+    ArrowAbiUtil.exportSchema(allocator, arrowSchema, cSchema)
+    val c2cHandle = jniWrapper.init(
+      cSchema.memoryAddress(),
+      NativeMemoryManagers
+        .contextInstance("VanillaColumnarToNativeColumnar")
+        .getNativeInstanceHandle)
+
+    val arrowConverter = ArrowColumnarBatchConverter.create(arrowSchema, allocator)
+
+    TaskResources.addRecycler("VanillaColumnarToNativeColumnar_resourceClean", 100) {
+      jniWrapper.close(c2cHandle)
+      arrowConverter.close()
+      cSchema.close()
+    }
+
+    val res: Iterator[ColumnarBatch] = new Iterator[ColumnarBatch] {
+
+      var arrowArray: ArrowArray = null
+      TaskResources.addRecycler("VanillaColumnarToNativeColumnar_arrowArray", 100) {
+        if (arrowArray != null) {
+          arrowArray.release()
+          arrowArray.close()
+          arrowConverter.reset()
+        }
+      }
+
+      override def hasNext: Boolean = {
+        if (arrowArray != null) {
+          arrowArray.release()
+          arrowArray.close()
+          arrowConverter.reset()
+          arrowArray = null
+        }
```

Comment:
Seems to be that the file includes a bunch of manual life cycle management calls like `_.close` `_.release` `_.reset`, etc. Would you like to review them to see if we can eliminate some? I recall that `recycleIterator` / `recyclePayloads` can do some of the clean up jobs automatically.

It's a non-blocking comment so feel free to do that in another PR. But anyway we'd try best to make code here easier to maintain from further developers.

Reply:
We need to reset arrowArray and arrowConverter when iterate each item, so `recyclePayload` is reasonable, but it only passes the result of the `next()`, which is `ColumnarBatch`, we cannot pass arrowArray to close them.

So here instead I add an extra method `releaseArrowBuf` to release them to reduce duplicates.